serverless/pytorch/mmpose/ubody2d/nuclio/main.py

```diff
@@ -0,0 +1,112 @@
+import json
+import base64
+import io
+import yaml
+import numpy as np
+from PIL import Image
+
+from mmpose.apis import inference_topdown
+from mmpose.apis import init_model as init_pose_estimator
+from mmdet.apis import inference_detector, init_detector
+from mmpose.evaluation.functional import nms
+from mmpose.structures import merge_data_samples
+from mmpose.utils import adapt_mmdet_pipeline
+
+
+def process_one_image(img, detector, pose_estimator, threshold):
+    category_ids = [0]
+    # predict bbox
+    det_result = inference_detector(detector, img)
+    pred_instance = det_result.pred_instances.cpu().numpy()
+    bboxes = np.concatenate(
+        (pred_instance.bboxes, pred_instance.scores[:, None]), axis=1)
+    bboxes = bboxes[np.logical_and(pred_instance.labels == category_ids,
+                                   pred_instance.scores > threshold)]
+    bboxes = bboxes[nms(bboxes, threshold), :4]
+
+    # predict keypoints
+    pose_results = inference_topdown(pose_estimator, img, bboxes)
+    data_samples = merge_data_samples(pose_results)
+
+    # if there is no instance detected, return None
+    return data_samples.get('pred_instances', None)
+
+
+def init_context(context):
+    det_config = '/opt/nuclio/mmpose/projects/rtmpose/rtmdet/person/rtmdet_nano_320-8xb32_coco-person.py'
+    det_checkpoint = '/opt/nuclio/rtmdet_nano_8xb32-100e_coco-obj365-person-05d8511e.pth'
+    pose_config = '/opt/nuclio/mmpose/configs/wholebody_2d_keypoint/topdown_heatmap/ubody2d/td-hm_hrnet-w32_8xb64-210e_ubody-256x192.py'
+    pose_checkpoint = '/opt/nuclio/td-hm_hrnet-w32_8xb64-210e_ubody-coco-256x192-7c227391_20230807.pth'
+
+    # build person detector
+    context.logger.info("Init detector...")
+    detector = init_detector(det_config, det_checkpoint, device='cpu')
+    detector.cfg = adapt_mmdet_pipeline(detector.cfg)
+
+    # build pose estimator
+    context.logger.info("Init pose estimator...")
+    pose_estimator = init_pose_estimator(pose_config, pose_checkpoint, device='cpu')
+
+    context.logger.info("Init labels...")
+    with open("/opt/nuclio/function.yaml", 'rb') as function_file:
+        functionconfig = yaml.safe_load(function_file)
+        labels_spec = functionconfig['metadata']['annotations']['spec']
+        labels = [{
+            "name": item["name"],
+            "elements": {
+                element["id"]: {
+                    "name": element["name"]
+                } for element in item["elements"]
```

Comment:
I'm not sure there's any point in converting `item["elements"]` to a dictionary. AFAICS, you only use this dictionary once, and in that one instance, you iterate over the whole thing anyway. So you might as well keep it as a list.

Reply:
Adjusted